api/net/Areas/Subscriber/Controllers/ReportController.cs

```diff
@@ -476,7 +476,56 @@ public async Task<IActionResult> AddContentToReportAsync(int id, [FromBody] IEnu
             var instances = _reportService.GetLatestInstances(id, user.Id);
             var currentInstance = instances.FirstOrDefault() ?? throw new InvalidOperationException("Unable to add content to a report without an instance");
             currentInstance.ContentManyToMany.Clear();
-            currentInstance.ContentManyToMany.AddRange(_reportInstanceService.GetContentForInstance(currentInstance.Id));
+            currentInstance.ContentManyToMany.AddRange(_reportInstanceService.GetContentForInstanceBasic(currentInstance.Id));
+            result.Instances.Clear();
+            result.Instances.AddRange(instances);
+
+            var ownerId = result.OwnerId ?? currentInstance.OwnerId;
+            if (ownerId.HasValue)
+            {
+                user = _userService.FindByIdMinimal(ownerId.Value) ?? throw new NotAuthorizedException();
+                await _kafkaMessenger.SendMessageAsync(
+                    _kafkaHubOptions.HubTopic,
+                    new KafkaHubMessage(HubEvent.SendUser, user.Username, new KafkaInvocationMessage(MessageTarget.ReportStatus, new[] { new ReportMessageModel(currentInstance) }))
+                );
+            }
+            return new JsonResult(new ReportModel(result, _serializerOptions));
+        }
+
+        return new JsonResult(new ReportModel(report, _serializerOptions));
+    }
+
+    /// <summary>
+    /// Fast path: Add specified content to the current report instance with minimal DB updates.
+    /// Avoids full instance graph merges for large reports.
+    /// </summary>
+    /// <param name="id"></param>
+    /// <param name="content"></param>
+    /// <returns></returns>
+    /// <exception cref="NotAuthorizedException"></exception>
+    /// <exception cref="NoContentException"></exception>
+    [HttpPost("{id}/content/fast")]
+    [Produces(MediaTypeNames.Application.Json)]
+    [ProducesResponseType(typeof(ReportModel), (int)HttpStatusCode.OK)]
+    [ProducesResponseType(typeof(ErrorResponseModel), (int)HttpStatusCode.BadRequest)]
+    [SwaggerOperation(Tags = new[] { "Report" })]
+    public async Task<IActionResult> FastAddContentToReportAsync(int id, [FromBody] IEnumerable<ReportInstanceContentModel> content)
+    {
+        var user = _impersonate.GetCurrentUser();
+        var report = _reportService.FindById(id) ?? throw new NoContentException("Report does not exist");
+        if (report.OwnerId != user.Id &&
+            !report.SubscribersManyToMany.Any(s => s.IsSubscribed && s.UserId == user.Id) &&
```

Comment:
This looks like a logic/bug in the original code.  I don't know why a subscriber of a report should be allowed to add content to someone else's report.